kong/router/compat.lua

```diff
@@ -295,8 +305,62 @@ local function get_exp_and_priority(route)
 end
 
 
-function _M.new(routes, cache, cache_neg, old_router)
-  return atc.new(routes, cache, cache_neg, old_router, get_exp_and_priority)
+-- group array-like table t by the function f, returning a table mapping from
+-- the result of invoking f on one of the elements to the actual elements.
+local function group_by(t, f)
+  local result = {}
+  for _, value in ipairs(t) do
+    local key = f(value)
+    if result[key] then
+      table.insert(result[key], value)
+    else
+      result[key] = {value}
+    end
+  end
+  return result
+end
+
+-- split routes into multiple routes, one for each prefix length and one for all
+-- regular expressions
+local function split_route_by_path_into(rs, split_rs)
+  if is_empty_field(rs.route.paths) or #rs.route.paths == 1 then
+    table.insert(split_rs, rs)
+    return
+  end
+
+  assert(tablex.size(rs) == 2) -- make sure that rs contains only the two expected entries, route and service
+
+  local grouped_paths = group_by(
+    rs.route.paths,
+    function(path)
+      return is_regex_magic(path) or #path
+    end
+  )
+  for _, paths in pairs(grouped_paths) do
+    local cloned_route = {
+      route = utils.shallow_copy(rs.route),
+      service = rs.service,
+    }
+    cloned_route.route.original_route = rs.route
+    cloned_route.route.paths = paths
+    cloned_route.route.id = utils.uuid()
```

Comment:
I am wondering what this would do to incremental rebuilds?

Reply:
I've changed the id generation for the split routes so that a uuidv5 is used, based on the original route ID and the index of the path.  That way, incremental rebuilds will not discard and rebuild these split routes if they were not changed.